DEVX-1263 DEVX-1269 Docs Conversion: secret protection beginner cloud (#767)
confluentinc/examples: ccloud/docs/beginner-cloud.rst

.. _beginner-cloud:

> Confluent Cloud CLI

Reviewer: ```suggestion
Tutorial: |ccloud| CLI
```

And updated `=====` underneath

Author: Updated

===================

Overview
--------

You can use `Confluent Cloud CLI
<https://docs.confluent.io/current/cloud/cli/install.html#ccloud-install-cli?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__
to interact with your `Confluent Cloud
<https://confluent.cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__
cluster.

> 

Reviewer: Need a bit more of an introduction, e.g. this tutorial will show you how ....

Author: Updated

> `start.sh <start.sh>`__ is a fully scripted demo that shows you how to interact

Reviewer: This paragraph and bullet points could go to `Run the Demo` section.

Author: Added

with |ccloud|. Using the |ccloud| CLI, you can complete the following workflow
in about 8 minutes:

-  Create a new environment and specify it as the default
-  Create a new |ak| cluster and specify it as the default
-  Create a user key/secret pair and specify it as the default
-  Produce and consume with |ccloud| CLI
-  Create a service account key/secret pair
-  Run a Java producer: before and after ACLs
-  Run a Java producer: showcase a Prefix ACL
-  Run |kconnect| and kafka-connect-datagen connector with permissions
-  Run a Java consumer: showcase a Wildcard ACL
-  Delete the API key, service account, |ak| topics, |ak| cluster, environment,
   and the log files

Prerequisites
~~~~~~~~~~~~~~

-  Access to a |ccloud| cluster

> -  Local install of `Confluent Cloud

Reviewer: Check link -- doesn't go to CLI page

Author: I will build and check

Author: Updated link

   v1.7.0 or later
-  |ccloud| user credentials saved in ``~/.netrc`` (save with command ``ccloud login --save``)
-  Docker and Docker Compose for the local |kconnect| worker
-  ``timeout`` installed on your host
-  ``mvn`` installed on your host
-  ``jq`` installed on your host

Confluent Cloud Promo Code
~~~~~~~~~~~~~~~~~~~~~~~~~~

The first 20 users to sign up for `Confluent Cloud
<https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__
and use promo code ``C50INTEG`` will receive an additional $50 free usage
(`details
<https://www.confluent.io/confluent-cloud-promo-disclaimer/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__).

Run the demo
------------

.. note::

   This example uses real resources in |ccloud|, and it creates and deletes
   topics, service accounts, API keys, and ACLs.

To run this demo, execute the following command:

.. code-block:: bash

      ./start.sh

> 

Reviewer: Proposal: let's capture the output of a successful run and walk through it here.  See DEVX-1263 for log output and suggestions.

Author: Got it

   * - ``CLUSTER_REGION``
     - us-west-2

Demo Cleanup
~~~~~~~~~~~~

If you run a demo that ends prematurely, you may receive the following error
message when trying to run the demo again (``ccloud environment create
demo-script-env``):

.. code-block:: text

      Error: 1 error occurred:
         * error creating account: Account name is already in use

      Failed to create environment demo-script-env. Please troubleshoot and run again

To perform demo cleanup, complete the following steps:

#. Delete the API keys and ACLs created in a previous demo run.

#. Run the following script to delete the demo’s topics, |ak| cluster, and environment.

   .. code-block:: bash

      ./cleanup.sh

Additional Resources
---------------------

-  See the `Best Practices for Developing Kafka Applications on
   Confluent Cloud
   <https://assets.confluent.io/m/14397e757459a58d/original/20200205-WP-Best_Practices_for_Developing_Apache_Kafka_Applications_on_Confluent_Cloud.pdf?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.ccloud>`__
   whitepaper for a guide to configuring, monitoring, and optimizing
   your |ak| client applications when using |ccloud|.

> -  See other `Confluent Cloud demos <../README.md>`__.

Author: @ybyzek Is this link correct?

Reviewer: Good catch.  This should be as follows:

```suggestion
-  See other :ref:`ccloud-demos-overview`.
```

Author: Updated

.. _beginner-cloud:

Tuturial: |ccloud| CLI
=======================

Overview
--------

> This tuturial will show you how to use the `Confluent Cloud CLI

Author: Is this intro okay, or we need a little more meat?

Reviewer: FYC: add the bulleted list of steps that is covered 

```
* Create a new environment and specify it as the default
 * Create a new Kafka cluster and specify it as the default		
 * Create a user key/secret pair and specify it as the default		
 * Produce and consume with Confluent Cloud CLI		
 * Create a service account key/secret pair		
 * Run a Java producer: before and after ACLs		
 * Run a Java producer: showcase a Prefix ACL		
 * Run Connect and kafka-connect-datagen connector with permissions		
 * Run a Java consumer: showcase a Wildcard ACL		
 * Delete the API key, service account, Kafka topics, Kafka cluster, environment, and the log files
```

After running the previous command, you can complete the following workflow
in about 8 minutes using the |ccloud| CLI:

.. note::

   This example uses real resources in |ccloud|, and it creates and deletes
   topics, service accounts, API keys, and ACLs.

-  `Create a new environment and specify it as the default`_
-  `Create a new Kafka cluster and specify it as the default`_
-  `Create a user key/secret pair and specify it as the default`_
-  `Produce and consume with Confluent Cloud CLI`_
-  `Create a service account key/secret pair`_
-  `Run a Java producer: before and after ACLs`_
-  `Run a Java producer: showcase a Prefix ACL`_
-  `Run |kconnect| and kafka-connect-datagen connector with permissions`_
-  `Run a Java consumer: showcase a Wildcard ACL`_
-  `Delete the API key, service account, Kafka topics, Kafka cluster, environment,
   and the log files`_

Create a new environment and specify it as the default
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following command to create a new |ccloud| environment
   ``demo-script-env``:

   .. code-block:: bash

      ccloud environment create demo-script-env -o json

   The output is in JSON format:

> 

Author: I'm not too sure if this is all we want to say for the JSON data here.  Should there be a step for the user to view the JSON file and verify the data keys/values are set? @ybyzek. This also occurs in a few other spots in our page here. So what we decide here, I'll make the updates in the rest of the page accordingly

Reviewer: @tikimims How about one step to take action (e.g. `run the command...`) and one step to verify output (e.g. `verify that the output resembles...`)?  Note, there is not JSON "file", it's just command *output* formatted as JSON>

Author: Cool beans. Updated

   .. code-block:: text

      {
        "id": "env-5qz2q",
        "name": "demo-script-env"
      }

#. Specify ``env-5qz2q`` as the active environment by running the following
   command:

   .. code-block:: bash

       ccloud environment use env-5qz2q

> 

Author: Wanted to confirm here that the user does in fact see the following message after specifying the active environment? Wasn't sure if was a comment per se when looking at the log file write-up or if its an actual message the user sees

Reviewer: Confirmed -- it's an actual message.

Author: Cool beans

      cloud api-key create --description "Demo credentials" --resource lkc-x6m01 -o json

   The output is in JSON format:

   .. code-block:: text

      {
         "key": "QX7X4VA4DFJTTOIA",
         "secret": "fjcDDyr0Nm84zZr77ku/AQqCKQOOmb35Ql68HQnb60VuU+xLKiu/n2UNQ0WYXp/D"
      }

#. Associate the API key ``QX7X4VA4DFJTTOIA`` to the Kafka cluster ``lkc-x6m01``:

   .. code-block:: bash

      ccloud api-key use QX7X4VA4DFJTTOIA --resource lkc-x6m01

   Running the previoud command sets the API Key "QX7X4VA4DFJTTOIA" as the
   active API key for ``lkc-x6m0``.

#. Verify you see the following message:

   .. code-block:: text

      Waiting for Confluent Cloud cluster to be ready and for credentials to propagate
      ....

Produce and consume records with Confluent Cloud CLI
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Produce records
^^^^^^^^^^^^^^^

#. Run the following command to create a new Kafka topic ``demo-topic-1``:

   .. code-block:: bash

      ccloud kafka topic create demo-topic-1

#. Produce 10 messages to topic ``demo-topic-1`` by completing the following steps:

   a. Implement the following logic:

      .. code-block:: text

>          (for i in `seq 1 10`; do echo "${i}" ; done) | \ timeout 10s

Author: In this logic is the "timeout 10s"  meant to be formatted as a comment?

Reviewer: No, it's an actual bash command.

Author: Got it

   .. code-block:: text

      Starting Kafka Consumer. ^C or ^D to exit
      2
      3
      9
      4
      5
      7
      10
      1
      6
      8

Create a service account key/secret pair
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following commmand to create a new service account:

   .. code-block:: bash

      ccloud service-account create demo-app-3288 --description demo-app-3288 -o json

   .. code-block:: text

      {
         "id": 104349,
         "name": "demo-app-3288",
          "description": "demo-app-3288"
      }

#. Create an API key and secret for the service account ``104349`` by running
   the following account:

   .. code-block:: bash

      ccloud api-key create --service-account 104349 --resource lkc-x6m01 -o json

   .. code-block:: text

      {
        "key": "ESN5FSNDHOFFSUEV",
        "secret": "nzBEyC1k7zfLvVON3vhBMQrNRjJR7pdMc2WLVyyPscBhYHkMwP6VpPVDTqhctamB"
      }

#. Complete the following steps to create a local configuration file
   ``/tmp/client.config`` with |ccloud| connection information using the newly
   created API key and secret:

   a. Open the config file:

      cat /tmp/client.config

> 

Author: Is the following step correct?

.. _beginner-cloud:

Tuturial: |ccloud| CLI
=======================

Overview
--------

This tuturial will show you how to use the `Confluent Cloud CLI

> <https://docs.confluent.io/current/cloud/cli/install.html#ccloud-install-cli?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__

Reviewer: UTMs can be stripped.

Author: Link updated

to interact with your `Confluent Cloud

> <https://confluent.cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__

Reviewer: UTMs can be stripped

cluster.

Prerequisites
~~~~~~~~~~~~~~

> -  Access to a |ccloud| cluster

Reviewer: Add link to CCloud?

Author: Good one. Yep added.

-  Local install  `Confluent Cloud

>    CLI <https://docs.cofonfluent.io/current/cloud/cli/install.html#ccloud-install-cli?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__

Reviewer: UTMs can be stripped

   v1.7.0 or later
-  |ccloud| user credentials saved in ``~/.netrc`` (save with command ``ccloud login --save``)
-  Docker and Docker Compose for the local |kconnect| worker
-  ``timeout`` installed on your host
-  ``mvn`` installed on your host
-  ``jq`` installed on your host

Confluent Cloud Promo Code
~~~~~~~~~~~~~~~~~~~~~~~~~~

The first 20 users to sign up for `Confluent Cloud
<https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__
and use promo code ``C50INTEG`` will receive an additional $50 free usage
(`details
<https://www.confluent.io/confluent-cloud-promo-disclaimer/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.beginner-cloud>`__).

Run the demo
------------

> `start.sh <start.sh>`__ is a fully scripted demo that shows you how to interact

Reviewer: This link won't work without `:devx-examples:`...

with |ccloud|.

To run the demo, execute the following command:

.. code-block:: bash

      ./start.sh

After running the previous command, you can complete the following workflow
in about 8 minutes using the |ccloud| CLI:

.. note::

>    This example uses real resources in |ccloud|, and it creates and deletes

Reviewer: FYC: Maybe take this out of the `note` and up into the overview section?

   topics, service accounts, API keys, and ACLs.

-  `Create a new environment and specify it as the default`_
-  `Create a new Kafka cluster and specify it as the default`_
-  `Create a user key/secret pair and specify it as the default`_
-  `Produce and consume with Confluent Cloud CLI`_
-  `Create a service account key/secret pair`_
-  `Run a Java producer: before and after ACLs`_
-  `Run a Java producer: showcase a Prefix ACL`_
-  `Run |kconnect| and kafka-connect-datagen connector with permissions`_
-  `Run a Java consumer: showcase a Wildcard ACL`_
-  `Delete the API key, service account, Kafka topics, Kafka cluster, environment,
   and the log files`_

Create a new environment and specify it as the default
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following command to create a new |ccloud| environment
   ``demo-script-env``:

   .. code-block:: bash

      ccloud environment create demo-script-env -o json

   The output is in JSON format:

   .. code-block:: text

      {
        "id": "env-5qz2q",
        "name": "demo-script-env"
      }

> #. Specify ``env-5qz2q`` as the active environment by running the following

Reviewer: So this is interesting.  We should tell the user to observe that the environment id ``env-5qz2q`` was in the previous output, and we are referencing it in the next command.

Author: Yes I thought something was quirky there. Will update

   command:

   .. code-block:: bash

       ccloud environment use env-5qz2q

   You should see the following message: ``Now using "env-5qz2q" as the default (active) environment.``

Create a new Kafka cluster and specify it as the default
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following command to create a new |ccloud| cluster

>    ``demo-kafka-cluster``:

Reviewer: Let's tell the user that they can choose whatever provider/region they want, per the output of ``ccloud kafka region list``

Author: Updated

   .. code-block:: bash

      ccloud kafka cluster create demo-kafka-cluster --cloud aws --region us-west-2

   .. note::

>       It may take up to 5 minutes for the |ak| cluster to be ready.

Reviewer: My leaning is that `note` is distracting, so it's better to integrate the text into the step above.

Author: Updated

> #. Verify you see the following output:

Reviewer: It won't be exact...it will resemble ;)

   .. code-block:: text

      +--------------+---------------------------------------------------------+
      | Id           | lkc-x6m01                                               |
      | Name         | demo-kafka-cluster                                      |
      | Type         | BASIC                                                   |
      | Ingress      |                                                     100 |
      | Egress       |                                                     100 |
      | Storage      |                                                    5000 |
      | Provider     | aws                                                     |
      | Availability | LOW                                                     |
      | Region       | us-west-2                                               |
      | Status       | UP                                                      |
      | Endpoint     | SASL_SSL://pkc-4kgmg.us-west-2.aws.confluent.cloud:9092 |
      | ApiEndpoint  | https://pkac-ldgj1.us-west-2.aws.confluent.cloud        |
      +--------------+---------------------------------------------------------+

> #. Specify ``lkc-x6m01`` as the active |ak|| cluster by running the following

Reviewer: ```suggestion
#. Specify ``lkc-x6m01`` as the active |ak| cluster by running the following
```

Author: Done

   command:

   .. code-block:: bash

      ccloud kafka cluster use lkc-x6m01

>    Running the previous command sets the |ak| cluster ``lkc-x6m01`` as the active

Reviewer: Can this statement be integrated into the step text itself?

Author: Done

   cluster for environment ``env-5qz2``.

Create a user key/secret pair and specify it as the default
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following command to create a user key/secret pair:

   .. code-block:: bash

      cloud api-key create --description "Demo credentials" --resource lkc-x6m01 -o json

   The output is in JSON format:

   .. code-block:: text

      {
         "key": "QX7X4VA4DFJTTOIA",
         "secret": "fjcDDyr0Nm84zZr77ku/AQqCKQOOmb35Ql68HQnb60VuU+xLKiu/n2UNQ0WYXp/D"
      }

> #. Associate the API key ``QX7X4VA4DFJTTOIA`` to the Kafka cluster ``lkc-x6m01``:

Reviewer: ```suggestion
#. Associate the API key ``QX7X4VA4DFJTTOIA`` to the |ak| cluster ``lkc-x6m01``:
```

Author: Done

   .. code-block:: bash

      ccloud api-key use QX7X4VA4DFJTTOIA --resource lkc-x6m01

   Running the previoud command sets the API Key "QX7X4VA4DFJTTOIA" as the
   active API key for ``lkc-x6m0``.

> #. Verify you see the following message:

Reviewer: Actually this is from the script, not from the command.  I think we need to remove this but tell the user something like "Please wait 90 seconds for Confluent Cloud cluster to be ready and for credentials to propagate"

Author: Got it

   .. code-block:: text

      Waiting for Confluent Cloud cluster to be ready and for credentials to propagate
      ....

Produce and consume records with Confluent Cloud CLI
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Produce records
^^^^^^^^^^^^^^^

#. Run the following command to create a new Kafka topic ``demo-topic-1``:

   .. code-block:: bash

      ccloud kafka topic create demo-topic-1

#. Produce 10 messages to topic ``demo-topic-1`` by completing the following steps:

   a. Implement the following logic:

      .. code-block:: text

         (for i in `seq 1 10`; do echo "${i}" ; done) | \ timeout 10s

   b. Run the following command:

      .. code-block:: bash

         ccloud kafka topic produce demo-topic-1

#. Verify you see the following output:

   .. code-block:: text

      Starting Kafka Producer. ^C or ^D to exit
      1
      2
      3
      4
      5
      6
      7
      8
      9
      10

Consume records
^^^^^^^^^^^^^^^

> #. Run the following command to consume messages from topic ``demo-topic-1``:

Reviewer: Let's tell the user what the `-b` means.  (consume from beginning of topic)

Author: Updated

   .. code-block:: bash

      ccloud kafka topic consume demo-topic-1 -b

#. Verify you see the following output:

   .. code-block:: text

      Starting Kafka Consumer. ^C or ^D to exit
      2
      3
      9
      4
      5
      7
      10
      1
      6
      8

Create a service account key/secret pair
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following commmand to create a new service account:

   .. code-block:: bash

      ccloud service-account create demo-app-3288 --description demo-app-3288 -o json

> 

Reviewer: Should we have here `#. Verify your output resembles...`?

   .. code-block:: text

      {
         "id": 104349,
         "name": "demo-app-3288",
          "description": "demo-app-3288"
      }

#. Create an API key and secret for the service account ``104349`` by running
   the following account:

   .. code-block:: bash

      ccloud api-key create --service-account 104349 --resource lkc-x6m01 -o json

> 

Reviewer: Same...need text like `#. Verify your output resembles...`?

Author: Yes sounds good. Will update

   .. code-block:: text

      {
        "key": "ESN5FSNDHOFFSUEV",
        "secret": "nzBEyC1k7zfLvVON3vhBMQrNRjJR7pdMc2WLVyyPscBhYHkMwP6VpPVDTqhctamB"
      }

#. Complete the following steps to create a local configuration file
   ``/tmp/client.config`` with |ccloud| connection information using the newly
   created API key and secret:

   a. Open the config file:

      cat /tmp/client.config

   b. Update  ``/tmp/client.config``  with the following parameters:

      .. code-block:: text

         ssl.endpoint.identification.algorithm=https
         sasl.mechanism=PLAIN
         security.protocol=SASL_SSL
         bootstrap.servers=pkc-4kgmg.us-west-2.aws.confluent.cloud:9092
         sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username\="ESN5FSNDHOFFSUEV" password\="nzBEyC1k7zfLvVON3vhBMQrNRjJR7pdMc2WLVyyPscBhYHkMwP6VpPVDTqhctamB";

>    c. Wait 90 seconds for the service account credentials to propagate

Reviewer: These should be promoted to numbered steps, not (c), (d)

Author: Will update in New PR

   d. By default, no ACLs are configured. To confirm, run the following command:

      .. code-block:: bash

         ccloud kafka acl list --service-account 104349

      You should see the following output:

      .. code-block:: text

            ServiceAccountId | Permission | Operation | Resource | Name | Type
          +------------------+------------+-----------+----------+------+------+

Run a Java producer before and after configuring ACLs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the Java producer to ``demo-topic-1`` before configuring ACLs (expected to fail):

   .. code-block:: bash

      mvn -q -f ../../clients/cloud/java/pom.xml exec:java -Dexec.mainClass="io.confluent.examples.clients.cloud.ProducerExample" -Dexec.args="/tmp/client.config demo-topic-1" -Dlog4j.configuration=file:log4j.properties > /tmp/log.1 2>&1

#. Verify you see ``org.apache.kafka.common.errors.TopicAuthorizationException``

>    in the logs as shown in the following example (expected because there are no ACLs to allow this client application):

Reviewer: ```suggestion
   in the log file ``/tmp/log.1`` as shown in the following example (expected because there are no ACLs to allow this client application):
```

Author: updated in new PR

         ServiceAccountId | Permission | Operation | Resource |     Name     |  Type
       +------------------+------------+-----------+----------+--------------+---------+
         User:104349      | ALLOW      | WRITE     | TOPIC    | demo-topic-1 | LITERAL

#. Run the following command and verify the ACLs were configured:

   .. code-block:: bash

      ccloud kafka acl list --service-account 104349

   You should see:

   .. code-block:: text

         ServiceAccountId | Permission | Operation | Resource |     Name     |  Type
       +------------------+------------+-----------+----------+--------------+---------+
         User:104349      | ALLOW      | CREATE    | TOPIC    | demo-topic-1 | LITERAL
         User:104349      | ALLOW      | WRITE     | TOPIC    | demo-topic-1 | LITERAL

#. Run the Java producer to ``demo-topic-1`` after configuring the ACLs:

   .. code-block:: bash

      mvn -q -f ../../clients/cloud/java/pom.xml exec:java -Dexec.mainClass="io.confluent.examples.clients.cloud.ProducerExample" -Dexec.args="/tmp/client.config demo-topic-1" -Dlog4j.configuration=file:log4j.properties > /tmp/log.2 2>&1

> #. Verify you see the ``10 messages were produced to topic`` message in the logs as

Reviewer: ```suggestion
#. Verify you see the ``10 messages were produced to topic`` message in the log file ``/tmp/log.2``  as
```

Author: Updated in new PR

         Produced record to topic demo-topic-1 partition [3] @ offset 2
         Produced record to topic demo-topic-1 partition [3] @ offset 3
         Produced record to topic demo-topic-1 partition [3] @ offset 4
         Produced record to topic demo-topic-1 partition [3] @ offset 5
         Produced record to topic demo-topic-1 partition [3] @ offset 6
         Produced record to topic demo-topic-1 partition [3] @ offset 7
         Produced record to topic demo-topic-1 partition [3] @ offset 8
         Produced record to topic demo-topic-1 partition [3] @ offset 9
         10 messages were produced to topic demo-topic-1

#. Run the following commands to delete the ACLs:

   .. code-block:: bash

      ccloud kafka acl delete --allow --service-account 104349 --operation CREATE --topic demo-topic-1
      ccloud kafka acl delete --allow --service-account 104349 --operation WRITE --topic demo-topic-1

   You should see a ``Deleted ACLs`` message.

#. Create a new Kafka topic ``demo-topic-2``:

   .. code-block:: bash

      ccloud kafka topic create demo-topic-2

   You should see a ``Created topic "demo-topic-2"`` message.

> .. shouldn't the following say "prefixed ACL" instead of "prefix ACL"

Author: The following heading should say "prefixed ACL" instead of "prefix ACL?"

Reviewer: @tikimims great question, and we should be consistent with what the Confluent docs say -- can you please double check?

Author: Yep. Looks that way. I'll update it. Yesterday, I was reading over the following page to gain a little more understanding of ACL concepts. https://docs.confluent.io/current/kafka/authorization.html#prefixed-acls

Run a Java producer to showcase a prefix ACL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the following command to create ACLs for the producer using a prefix:

   .. code-block:: bash

> 

Author: .. should the commands say "--prefixed" instead of "--prefix?"

Reviewer: The command itself is `--prefix`

Author: Okay cool

      User:104349      | ALLOW      | CREATE    | TOPIC    | demo-topic | PREFIXED

      ServiceAccountId | Permission | Operation | Resource |    Name    |   Type
      +------------------+------------+-----------+----------+------------+----------+
      User:104349      | ALLOW      | WRITE     | TOPIC    | demo-topic | PREFIXED

#. Verify the ACLs were configured by running the following command:

   .. code-block:: bash

      ccloud kafka acl list --service-account 104349

   .. code-block:: text

         ServiceAccountId | Permission | Operation | Resource |    Name    |   Type
       +------------------+------------+-----------+----------+------------+----------+
         User:104349      | ALLOW      | WRITE     | TOPIC    | demo-topic | PREFIXED
         User:104349      | ALLOW      | CREATE    | TOPIC    | demo-topic | PREFIXED

#. Run the Java producer to ``demo-topic-2`` to prefix the ACLs:

   .. code-block:: bash

      mvn -q -f ../../clients/cloud/java/pom.xml exec:java -Dexec.mainClass="io.confluent.examples.clients.cloud.ProducerExample" -Dexec.args="/tmp/client.config demo-topic-2" -Dlog4j.configuration=file:log4j.properties > /tmp/log.3 2>&1

> #. Verify you see the ``10 messages were produced to topic`` message in the logs as

Reviewer: ```suggestion
#. Verify you see the ``10 messages were produced to topic`` message in the log file ``/tmp/log.3``  as
```

Author: updated in new PR

      Producing record: alice	{"count":8}
      Producing record: alice	{"count":9}
      Produced record to topic demo-topic-2 partition [3] @ offset 0
      Produced record to topic demo-topic-2 partition [3] @ offset 1
      Produced record to topic demo-topic-2 partition [3] @ offset 2
      Produced record to topic demo-topic-2 partition [3] @ offset 3
      Produced record to topic demo-topic-2 partition [3] @ offset 4
      Produced record to topic demo-topic-2 partition [3] @ offset 5
      Produced record to topic demo-topic-2 partition [3] @ offset 6
      Produced record to topic demo-topic-2 partition [3] @ offset 7
      Produced record to topic demo-topic-2 partition [3] @ offset 8
      Produced record to topic demo-topic-2 partition [3] @ offset 9
      10 messages were produced to topic demo-topic-2

#. Run the following commands to delete ACLs:

   .. code-block:: bash

      ccloud kafka acl delete --allow --service-account 104349 --operation CREATE --topic demo-topic --prefix
      ccloud kafka acl delete --allow --service-account 104349 --operation WRITE --topic demo-topic --prefix

   You should see a ``Deleted ACLs`` message.

> #. Create a new Kafka topic ``demo-topic-3``:

Reviewer: This should be moved to the next section

Author: updated

       +------------------+------------+-----------+----------+------+---------+
         User:104349      | ALLOW      | CREATE    | TOPIC    | *    | LITERAL

#. Run the following command to allow user ``104349`` to write to any topic
   in the ACL you created previously:

   .. code-block:: bash

      ccloud kafka acl create --allow --service-account 104349 --operation WRITE --topic '*'

   You should see the following output:

   .. code-block:: text

         ServiceAccountId | Permission | Operation | Resource | Name |  Type
       +------------------+------------+-----------+----------+------+---------+
         User:104349      | ALLOW      | WRITE     | TOPIC    | *    | LITERAL

#. Run the following command to allow user ``104349`` to read from any topic
   in the ACL:

   .. code-block:: bash

      ccloud kafka acl create --allow --service-account 104349 --operation READ --topic '*'

   .. code-block:: text

         ServiceAccountId | Permission | Operation | Resource | Name |  Type
       +------------------+------------+-----------+----------+------+---------+
         User:104349      | ALLOW      | READ      | TOPIC    | *    | LITERAL

> #. Run the following command to allow user ``104349`` to read from any topic

Reviewer: ```suggestion
#. Run the following command to allow user ``104349`` to have a consumer group called `connect`
```

Author: Updated

       +------------------+------------+-----------+----------+---------+---------+
         User:104349      | ALLOW      | WRITE     | TOPIC    | *       | LITERAL
         User:104349      | ALLOW      | CREATE    | TOPIC    | *       | LITERAL
         User:104349      | ALLOW      | READ      | TOPIC    | *       | LITERAL
         User:104349      | ALLOW      | READ      | GROUP    | connect | LITERAL

#. Generate environment variables with |ccloud| connection information for
   |kconnect| to use:

   .. code-block:: text

      ../../ccloud/ccloud-generate-cp-configs.sh /tmp/client.config &>/dev/null
      source delta_configs/env.delta

# Run a |kconnect| container with the kafka-connect-datagen plugin:

  .. code-block:: bash

     docker-compose up -d

  You should see the following output:

  .. code-block:: text

      Creating connect-cloud ... done
      Waiting up to 60 seconds for Docker container for connect to be up
      ............

#. Post the configuration for the kafka-connect-datagen connector that produces
   pageviews data to |ccloud| topic ``demo-topic-3``:

   .. code-block:: bash

>       curl --silent --output /dev/null -X POST -H "Content-Type: application/json" --data

Reviewer: Looks like the command was truncated.  Should be:

```
DATA=$( cat << EOF
{ 
  "name": "$CONNECTOR",
  "config": {
    "connector.class": "io.confluent.kafka.connect.datagen.DatagenConnector",
    "kafka.topic": "$TOPIC3",
    "quickstart": "pageviews",
    "key.converter": "org.apache.kafka.connect.storage.StringConverter",
    "value.converter": "org.apache.kafka.connect.json.JsonConverter",
    "value.converter.schemas.enable": "false",
    "max.interval": 5000,
    "iterations": 1000,
    "tasks.max": "1"
  }
}
EOF
)
curl --silent --output /dev/null -X POST -H "Content-Type: application/json" --data "${DATA}" http://localhost:8083/connectors
```

Author: Updated

   .. code-block:: text

      "{
      "name": "datagen-demo-topic-3",
      "config": {
         "connector.class": "io.confluent.kafka.connect.datagen.DatagenConnector",
         "kafka.topic": "demo-topic-3",
         "quickstart": "pageviews",
         "key.converter": "org.apache.kafka.connect.storage.StringConverter",
         "value.converter": "org.apache.kafka.connect.json.JsonConverter",
         "value.converter.schemas.enable": "false",
         "max.interval": 5000,
         "iterations": 1000,
         "tasks.max": "1"
      }
      }" http://localhost:8083/connectors

#. Wait 20 seconds for kafka-connect-datagen to start producing messages.

#. Run the following command to verify connector is running:

   .. code-block:: bash

>       curl --silent http://localhost:8083/connectors/datagen-demo-topic-3/status

Reviewer: ```suggestion
      curl --silent http://localhost:8083/connectors/datagen-demo-topic-3/status | jq -r '.'
```

Author: Updated

   .. code-block:: text

>       {"name":"datagen-demo-topic-3","connector":

Reviewer: Let's change the output to:

```
{
  "name": "datagen-demo-topic-3",
  "connector": {
    "state": "RUNNING",
    "worker_id": "connect:8083"
  },
  "tasks": [
    {
      "id": 0,
      "state": "RUNNING",
      "worker_id": "connect:8083"
    }
  ],
  "type": "source"
}
```

Author: Updated

      {"state":"RUNNING","worker_id":"connect:8083"},"tasks":
      [{"id":0,"state":"RUNNING","worker_id":"connect:8083"}],"type":"source"}

> #. Create ACLs for the consumer using a wildcard by running the following

Reviewer: This should be moved to the next section

Author: Updated

   You should see:

   .. code-block:: text

         ServiceAccountId | Permission | Operation | Resource |         Name          |  Type
       +------------------+------------+-----------+----------+-----------------------+---------+
         User:104349      | ALLOW      | READ      | GROUP    | connect               | LITERAL
         User:104349      | ALLOW      | CREATE    | TOPIC    | *                     | LITERAL
         User:104349      | ALLOW      | WRITE     | TOPIC    | *                     | LITERAL
         User:104349      | ALLOW      | READ      | TOPIC    | *                     | LITERAL
         User:104349      | ALLOW      | READ      | GROUP    | demo-beginner-cloud-1 | LITERAL

Run a Java consumer to showcase a Wildcard ACL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Run the Java consumer from ``demo-topic-3 (populated by kafka-connect-datagen)``:

   .. code-block:: bash

      mvn -q -f ../../clients/cloud/java/pom.xml exec:java -Dexec.mainClass="io.confluent.examples.clients.cloud.ConsumerExamplePageviews" -Dexec.args="/tmp/client.config demo-topic-3" -Dlog4j.configuration=file:log4j.properties > /tmp/log.4 2>&1

> #. Verify you see the ``Consumed record with`` message in the logs as

Reviewer: ```suggestion
#. Verify you see the ``Consumed record with`` message in the log file ``/tmp/log.4`` as
```

Author: Updated

Overview
--------

This tutorial will show you how to use the `Confluent Cloud CLI
<https://docs.confluent.io/current/cloud/cli/install.html>`__ to interact with
your `Confluent Cloud <https://confluent.cloud/login>`__ cluster.

> After running `start.sh <start.sh>`__, you can complete the following workflow

Reviewer: The following bullet list describes what start.sh does, not what is to be completed afterwards

Author: My bad – my moving things around made me lose the original purpose for that line. How about: 

> With `start.sh <start.sh>`__, a fully scripted demo that shows you how to interact
with |ccloud|, you can complete the following workflow in about 8
> minutes using the |ccloud| CLI:

Reviewer: Awesome!

Author: Sweet!! Thank you!

Author: Updated
